clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,83 @@
+use crate::utils::{is_expn_of, match_qpath, span_note_and_lint};
+use rustc::hir::{Expr, ExprKind, Mutability, UnOp};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function calls that mutate an object in `debug_assert!`, `debug_assert_eq!` and
+    /// `debug_assert_ne!` macros.
+    ///
+    /// **Why is this bad?** In release builds the code won't be executed and therefore can alter
+    /// the execution flow.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// debug_assert_eq!(vec![3].pop(), Some(3));
+    /// // or
+    /// let mut s = String::from("foo");
+    /// debug_assert_eq!(s.remove(0), 'f');
+    /// ```
+    pub DEBUG_ASSERT_WITH_MUT_CALL,
+    correctness,
+    "`debug_assert!` and others will be ignored in release mode and therefore alter the execution flow. Replace it with non-mut function calls."
+}
+
+declare_lint_pass!(DebugAssertWithMutCall => [DEBUG_ASSERT_WITH_MUT_CALL]);
+
+fn check_for_mutable_call(e: &Expr) -> bool {
+    match &e.kind {
+        ExprKind::Call(func, args) => {
+            if let ExprKind::Path(path) = &func.kind {
+                if !match_qpath(path, &["new_v1"]) {
+                    for a in (*args).iter() {
+                        if let ExprKind::AddrOf(Mutability::MutMutable, _) = &a.kind {
+                            return true;
+                        }
+                    }
+                }
+            }
+        },
+        _ => (),
+    }
+
+    false
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for DebugAssertWithMutCall {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, e: &'tcx Expr) {
+        fn check_inner(e: &Expr) -> Option<&Expr> {
+            if let ExprKind::AddrOf(_, addrof) = &e.kind {
+                if check_for_mutable_call(&addrof) {
+                    return Some(addrof);
+                }
+            }
+            None
+        }
+
+        let mut emit_lint: Option<&Expr> = None;
+        if is_expn_of(e.span, "debug_assert").is_some() {
```

Comment:
Maybe `is_direct_expn_of` is better here. (This may also make the `new_v1` check above useless) 

Maybe this lint is easier implemented, by first matching if the `expr` is a `{Method}Call` and the checking if it's a `is_direct_expn_of`.